chore(dependency)

tobiaskneubuehler opened the pull request:
- update dependencies (mobx still 4.x) ... add fix f…or proper matrix rendering because we have a timing-issue

- we have to update the frontend-logic dependency as soon as we release the updated version

fxstr commented on App.js:
```diff
@@ -68,8 +65,11 @@ export default class App extends React.Component {
     }
 
-    componentDidMount() {
+    async componentDidMount() {
```
What exactly happens if we async a componentDidMount? Will react delay rendering the component until the promise has been resolved? Will it wait with calling other lifecycle methods?

tobiaskneubuehler replied:
componentDidMount is the last lifecycle and happens after the render() method:
- so it does not block other lifecycle methods
- it does not block the rendering

fxstr commented on App.js:
```diff
@@ -68,8 +65,11 @@ export default class App extends React.Component {
         log('StatusBar:', StatusBar.curentHeight);
+
+        this.setupModelStateWatchers();
+        await this.setupApp();
```
What is the gain of calling this.setupApp here (instead of in the constructor)? In the constructor, the API call (that is essential to render the app) might be made a bit earlier, which IMO is advantageous.

tobiaskneubuehler replied:
we never call async functions in the constructor, right? they return a promise and the constructor is never asynchronous instead returns a instance of the class. @eventEmitter has a strong opinion on that

eventEmitter replied:
A really strong one indeed :dancer: The main file of an application _may_ be the only exception to this rule (does not apply to node.js).

TLDR: never hide asynchronous functionality to the outside.

In this case one could argue there isn't any need for exposing an async interface to the outside. The main file _may_ remove all interfaces to the outside, since they may not be required.

Since you are handling all errors anyway in the `setupApp` method, an outside party won't be able to handle them. The asynchronous nature of that method allows an outside party to wait for the app to be loaded though. You could also handle the errors on the outside of the class, or return re-throw them at least, which would allow full access and controllability to the loading process from the outside. 

As I once again don't know how the react thingy works, I cannot say with certainty what's right. 

You should probably decide if outside access is required. If not, you may handle the asynchronous stuff in the same stack as the constructor (I would prefer a separate method though).

```javascript
setupApp() {
        this.app.initialize().catch((err) => {
            this.app.errorHandler.handle(err);
            log('Error initializing app', err);
        }).finally(() => {
            log('App: Initialized');
        });
    }
```

In node applications, I try not to hide asynchronous things from the outside, since it has disadvantages, like for testing. My applications look normally like the dummy below (which I think is the right way to do things)

```javascript
// Application.js
class Application {
    constructor() {
        // set up the class, but don't ever do asyn stuff or event business logic
    }

     async load() {
         // do the asyn stuff and your business stuff
     }
}


// main.js
import Application from './Application.js';

const app = new Application();

app.load().then(() => {
     log('up\'n running');
}).catch(log);

// integration-test.js
import Application from './Application.js';

runTest(async () => {
     const app = new Application();
     await app.load();
     
     // wild tests!
     await app.end();
});
```

fxstr replied:
That is a very nice and understandable explication. I completely share your view. Therefore I stay with my argument:
1. `componentDidMount` is not the right place to load data in this case, as it is a React method that will never be `await`ed for
2. As all data is loaded into a MobX store that will trigger UI updates when changed, we should load data as **early** as possible.

As App.js **is** a React Native app's main file (as in Node's main.js) there's IMO no case where outside access could be needed. Therefore I think – in this specific case – App.js' constructor is the right place to call setupApp.

None the less, the whole App.js file could have been/should become cleaner.

fxstr commented on App.js:
```diff
@@ -143,7 +143,6 @@ export default class App extends React.Component {
     }
 
-    @observer
     render() {
```
Very good catch, thank you so, so much for this!

tobiaskneubuehler replied:
👍

tobiaskneubuehler commented on package.json:
```diff
@@ -11,20 +11,22 @@
-    "mobx-react": "^5.4.3",
-    "react": "16.5.0",
-    "react-native": "https://github.com/expo/react-native/archive/sdk-31.0.1.tar.gz",
+    "expo-codemod": "^1.0.0",
```
right, I may have user the `--save` parameter on `npm i` ... removed

tobiaskneubuehler commented on src/components/matrixContent/MatrixContent.js:
```diff
@@ -90,7 +90,22 @@ export default class MatrixContent extends React.Component {
     animatedBacteria = new Map();
 
-
+    /**
```
👍

tobiaskneubuehler commented on src/components/matrixContent/MatrixContent.js:
```diff
@@ -475,7 +511,16 @@ export default class MatrixContent extends React.Component {
 
                 { /* SUBSTANCE CLASSES (headers and lines) */ }
-                { this.props.matrix.defaultRadius &&
+                {/**
```
👍

tobiaskneubuehler commented on src/components/matrixContent/MatrixContent.js:
```diff
@@ -258,6 +273,27 @@ export default class MatrixContent extends React.Component {
             },
         );
 
+        /**
```
👍

tobiaskneubuehler commented on src/components/matrixContent/MatrixContent.js:
```diff
@@ -536,7 +581,16 @@ export default class MatrixContent extends React.Component {
                 { /* Container within which resistances will be moved/zoomed. Needed to
                      set the stage (container) and calculate its size for PanPinch */ }
-                { this.props.matrix.defaultRadius &&
+                {/**
```
👍

tobiaskneubuehler commented:
@fxstr I moved the timeout to the other reaction but let the action as we never change observables without an action, right?

tobiaskneubuehler commented:
oh, sry, didn't see the `runInAction` :/

fxstr commented:
> @fxstr I moved the timeout to the other reaction but let the action as we never change observables without an action, right?

Right. That's why I used a runInAction command:
``
`
setTimeout(() => {
    runInAction(() => { this.defaultRadiusWasSetDelayed = true; });
}, 0);
````

tobiaskneubuehler commented:
> > @fxstr I moved the timeout to the other reaction but let the action as we never change observables without an action, right?
> 
> Right. That's why I used a runInAction command:
> 
> ```
> setTimeout(() => {
>     runInAction(() => { this.defaultRadiusWasSetDelayed = true; });
> }, 0);
> ```
> 
> Both solutions are fine, one is a bit more verbous, the other one more dense.

yes, I missed that one ... changed to runInAction
